File: Crio/crio_programming_interview_problems-master/SymmetricTree/SymmetricTree.cpp
/*
class ListNode {
    public:
	int val;
	ListNode* next;

	ListNode() {}

	ListNode(int x) {                      
	    val = x;                       
	    next = NULL;
	}
};
*/
#include <bits/stdc++.h>
#include "FastIO.hpp"
#include "ReadMatrix.hpp"
#include "PrintMatrix.hpp"
#include "ListNode.hpp"
#include "TreeNode.hpp"

using namespace std;

class SymmetricTree {
public:
	// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    bool check(TreeNode* p,TreeNode* q)
    {
        if(p==NULL && q==NULL) return true;
        if(p==NULL || q==NULL) return false;
        if(p->val!=q->val) return false;
        if(!check(p->left,q->right)) return false;
        if(!check(p->right,q->left)) return false;
        return true;
    }
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

    bool isSymmetric(TreeNode* root) {
    	// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        if(root==NULL) return true;
        if(check(root->left,root->right)) return true;
        return false;
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
    }
};

int main() {
	FastIO();
	string s, num;

	getline(cin, s);
	vector<string> tree;
	stringstream Str(s);
	while(getline(Str, num, ' ')) {
		tree.push_back(num);
	}

	TreeNode* root = NULL;
	root = TreeNode().makeTreeFromArray(0, tree, root);
    bool result = SymmetricTree().isSymmetric(root);
    cout << boolalpha << result;
	return 0;
}

File: Crio/crio_programming_interview_problems-master/SymmetricTree/SymmetricTree.java
import crio.ds.Tree.TreeNode;
import java.io.*;
import java.util.*;

public class SymmetricTree {
    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    public boolean isMirror(TreeNode t1, TreeNode t2) {
        if (t1 == null && t2 == null)
            return true;
        if (t1 == null || t2 == null)
            return false;
        return (t1.val == t2.val) && isMirror(t1.right, t2.left) && isMirror(t1.left, t2.right);
    }
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

    // Implement your solution by completing the below function
    public boolean isSymmetric(TreeNode root) {
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        return isMirror(root, root);
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

        // CRIO_UNCOMMENT_START_MODULE_L1_PROBLEMS
        //return false;
        // CRIO_UNCOMMENT_END_MODULE_L1_PROBLEMS
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String input = in.readLine();
        String[] treeElements = new String[0];

        if (input.contains(" "))
            treeElements = input.split(" ");
        else if (!input.isEmpty())
            treeElements = new String[] { input };

        TreeNode root = TreeNode.generateTree(treeElements);
        boolean result = new SymmetricTree().isSymmetric(root);
        System.out.print(String.valueOf(result));
    }
}